lib/dal/src/workspace_snapshot/graph/v4/socket/input.rs

```diff
@@ -0,0 +1,177 @@
+use petgraph::prelude::*;
+use si_events::{ulid::Ulid, ContentHash};
+
+use crate::{
+    socket::input::InputSocketError,
+    workspace_snapshot::{
+        graph::{
+            traits::socket::input::InputSocketExt, LineageId, WorkspaceSnapshotGraphResult,
+            WorkspaceSnapshotGraphV4,
+        },
+        node_weight::{InputSocketNodeWeight, NodeWeight, NodeWeightError},
+    },
+    AttributeValueId, ComponentId, EdgeWeight, EdgeWeightKind, EdgeWeightKindDiscriminants,
+    InputSocketId, NodeWeightDiscriminants, SchemaVariantId, SocketArity,
+};
+
+impl InputSocketExt for WorkspaceSnapshotGraphV4 {
+    fn new_input_socket(
+        &mut self,
+        schema_variant_id: SchemaVariantId,
+        input_socket_id: InputSocketId,
+        lineage_id: LineageId,
+        arity: SocketArity,
+        content_hash: ContentHash,
+    ) -> WorkspaceSnapshotGraphResult<InputSocketNodeWeight> {
```

Comment:
I wonder if these traits should have an associated error type, or their own error type. Adding InputSocketError to the WorkspaceSnapshotGraphError seems like we're coupling the graph to the things on it even more closely. Not blocking at all, and something we can look at later.

Reply:
Yeah, it really is reinforcing my thinking that we really need to rework how we have errors set up.